docs/7-algorithms-and-data-structures-part-1/hometask/homework.ts

```diff
@@ -0,0 +1,189 @@
+  addVertex(vertex: T): void;
+
+  addEdge(vertex1: T, vertex2: T, weight: number): void;
+
+  getVertices(): Map<T, Map<T, number>>;
+}
+
+class WeightedGraphImpl<T> implements WeightedGraph<T> {
+  vertices: Map<T, Map<T, number>> = new Map();
+
+  addVertex(vertex: T): void {
+    this.vertices.set(vertex, new Map());
+  }
+
+  addEdge(vertex1: T, vertex2: T, weight: number): void {
+    const vertex = this.vertices.get(vertex1);
+    vertex && vertex.set(vertex2, weight);
+  }
+
+  getVertices(): Map<T, Map<T, number>> {
+    return this.vertices;
+  }
+}
+
+// Vertex / Edge
+class Vertex<T> {
+  constructor(public value: T) {
+  }
+}
+
+const vertices = [
+  new Vertex('1'),
+  new Vertex('2'),
+  new Vertex('3'),
+  new Vertex('4'),
+  new Vertex('5'),
+];
+
+class Edge<T> {
+  weight: number;
+  from: Vertex<T>;
+  to: Vertex<T>;
+
+  constructor(from: Vertex<T>, to: Vertex<T>, weight: number) {
+    this.weight = weight;
+    this.from = from;
+    this.to = to;
+  }
+}
+
+const edges = [
+  new Edge(vertices[0], vertices[3], 3),
+  new Edge(vertices[0], vertices[1], 5),
+  new Edge(vertices[0], vertices[2], 4),
+  new Edge(vertices[1], vertices[3], 6),
+  new Edge(vertices[1], vertices[2], 5),
+];
+
+// WeightedGraphImpl usages
+const graph: WeightedGraph<Vertex<string>> = new WeightedGraphImpl<Vertex<string>>();
+vertices.forEach(verticle => graph.addVertex(verticle));
+edges.forEach(edge => graph.addEdge(edge.from, edge.to, edge.weight));
+
+// Dijkstra
+interface Path {
+  path: string[];
+  distance: number;
+}
+
+interface Dijkstra<T> {
+  findShortestPath(vertex1: T, vertex2: T): Path;
+
+  findAllShortestPaths(vertex: T): Record<string, Path>;
+}
+
+class DijkstraImpl<T> implements Dijkstra<T> {
+  private readonly graph: WeightedGraph<T>;
+
+  constructor(graph: WeightedGraph<T>) {
+    this.graph = graph;
+  }
+
+  calculatedShortestPath(vertex1: T, vertex2: T): Path {
+    const vertices = this.graph.getVertices();
+    const isVertexes = vertex1 instanceof Vertex<string> && vertex2 instanceof Vertex<string>
+    const result: Path = {path: [], distance: 0};
+    const remoteLinks: Set<T> = new Set();
+
+    vertices.forEach((value, key) => {
+      value.has(vertex1 || vertex2) && remoteLinks.add(key);
+    });
+
+    if (remoteLinks.size < 2) {
+      result.distance = Infinity;
+      return result;
+    }
+
+    remoteLinks.forEach((valueRemote) => {
+      const isVertex = isVertexes && valueRemote instanceof Vertex<string>
+      let iteratedDistance: number = 0;
+      iteratedDistance += vertices.get(valueRemote)?.get(vertex1)!;
+      iteratedDistance += vertices.get(valueRemote)?.get(vertex2)!;
+      if (!Number.isInteger(iteratedDistance)) {
+        result.distance = Infinity;
+        return result;
+      }
+
+      if (((!result.distance) || iteratedDistance < result.distance) && isVertex) {
+        result.distance = iteratedDistance;
+        result.path = [vertex1.value, valueRemote.value, vertex2.value];
```

Comment:
how it will work if we will have more than 3 vertices between vertex1 and vertex2? please create this graph and add test for that case